src/Compiler/ES/TailRec.idr

```diff
@@ -87,218 +123,233 @@ import Data.String
-        Just x => pure x
-
-makeFusionBranch :  Name
-                 -> List (Name, Nat)
-                 -> (Nat, FC, List Name, ImperativeStatement)
-                 -> (ImperativeExp, ImperativeStatement)
-makeFusionBranch fusionName functionsIdx (i, _, args, body) =
-    let newArgExp = map (\i => IEConstructorArg (cast i) (IEVar fusionArgsName))
-                        [1..length args]
-        bodyRepArgs = replaceNamesExpS (zip args newArgExp) body
-    in (IEConstructorTag $ Left $ cast i, replaceExpS rep bodyRepArgs)
-    where
-        rep : ImperativeExp -> Maybe ImperativeExp
-        rep (IEApp (IEVar n) arg_vals) =
-            case lookup n functionsIdx of
-                Nothing => Nothing
-                Just i => Just $ IEApp
-                          (IEVar fusionName)
-                          [IEConstructor (Left $ cast i) arg_vals]
-        rep _ = Nothing
-
-changeBodyToUseFusion :  Name
-                      -> (Nat, Name, FC, List Name, ImperativeStatement)
-                      -> ImperativeStatement
-changeBodyToUseFusion fusionName (i, n, (fc, args, _)) =
-    FunDecl fc n args (ReturnStatement $ IEApp (IEVar fusionName)
-                        [IEConstructor (Left $ cast i) (map IEVar args)])
-
-tailRecOptimGroup :  {auto c : Ref TailRecS TailSt}
-                  -> SortedMap Name (FC, List Name, ImperativeStatement)
-                  -> List Name -> Core ImperativeStatement
-tailRecOptimGroup defs [] = pure neutral
-tailRecOptimGroup defs [n] =
-    do (fc, args , body) <- getDef defs n
-       pure $ FunDecl fc n args (makeTailOptimToBody n args body)
-
-tailRecOptimGroup defs names =
-    do fusionName <- genName
-       d <- traverse (getDef defs) names
-       let ids = [0..(length names `minus` 1)]
-       let alts = map (makeFusionBranch fusionName (zip names ids)) (zip ids d)
-       let fusionBody = SwitchStatement
-                          (IEConstructorHead $ IEVar fusionArgsName)
-                          alts
-                          Nothing
-       let fusionArgs = [fusionArgsName]
-       let fusion = FunDecl EmptyFC
-                            fusionName
-                            fusionArgs
-                            (makeTailOptimToBody fusionName fusionArgs fusionBody)
-       let newFunctions = Prelude.concat $ map
-                           (changeBodyToUseFusion fusionName)
-                           (ids `zip` (names `zip` d))
-       pure $ fusion <+> newFunctions
-
-||| (Mutual) Tail recursion optimizations: Converts all groups of
-||| mutually tail recursive functions to an imperative loop.
+
+--------------------------------------------------------------------------------
+--          Utilities
+--------------------------------------------------------------------------------
+
+-- indices of a list starting at 1
+indices : List a -> List Int
+indices = go 1
+  where go : Int -> List a -> List Int
+        go _ []       = []
+        go i (_ :: t) = i :: go (i+1) t
```

Comment:
```suggestion
indices : List a -> List Int
indices as = [1..length as]
```

Reply:
The types were wrong here (we need `Int` eventually, in a call to `MN`, that's why I didn't use `Nat` here), but I refactored this as suggested in a separate commit.